### PR TITLE
Refactor of `MultiheadAttention` module in PyTorch WMT workload

### PR DESCRIPTION
**Pull request by runame**
This is supposed to help with #467. I refactored the `MultiheadAttention` module to remove a lot of unnecessary code and fused the parameter matrices for the linear input projection, which leads to a slight improvement in speed.

**Timing results**
(note that all timings but the one of this refactor alone are just run once and there is probably a significant variance over multiple runs)

Original score (Juhan's timing): 4379
Score with this refactor: 4196/4253/4265
Score with `torch.compile` (Juhan's timing): 4236
Score with this refactor and `torch.compile`: 4163
Score with `torch.compile` and nightly (Juhan's timing): 4161
Score with this refactor, `torch.compile`, and nightly: error -> the last validation set batch size is different which causes this error. Can easily be fixed but since we are going to stick to stable for now I would defer making any changes until we decide to upgrade.

Clearly, all improvements are very minor and are not enough to bridge the gap to Jax.

@pomonam Can you please review my refactor of the `MultiheadAttention` module?
@znado Can you please review my changes to the `test_param_types` test?

*Note: This breaks the traindiffs code for WMT, will be fixed in a follow up PR.*

### CODE REVIEW

**Review by pomonam — DISMISSED** [review]
LGTM - thank you for fixing this! (It seems like every test is working now?)

**Comment by runame** [comment]
> It seems like every test is working now?

None of the GitHub tests were broken, the traindiffs test has to be run manually. @chandramouli-sastry already has a patch ready for after this PR is merged.

**Comment by runame** [comment]
I have added `torch.cuda.synchronize()` to the profiler (see latest commit). @pomonam Is there any reason not to do this?

**Comment by pomonam** [comment]
@runame Thank you! No, it was an oversight.

**Review by znado — APPROVED** [review]
